src/java.net.http/share/classes/jdk/internal/net/http/RequestPublishers.java

```diff
@@ -418,6 +419,90 @@ public long contentLength() {
         }
     }
 
+    public static final class FileChannelPublisher implements BodyPublisher {
+
+        private final FileChannel channel;
+
+        private final long position;
+
+        private final long limit;
+
+        public FileChannelPublisher(FileChannel channel, long offset, long length) {
+            this.channel = Objects.requireNonNull(channel, "channel");
+            long fileSize = fileSize(channel);
+            Objects.checkFromIndexSize(offset, length, fileSize);
```

Comment:
My preference would be to move these file size determination and the offset/length bounds check to the call site of this constructor. That way it's much more visible at the place where this behaviour is specified (by the public API).